hudi-io/src/main/java/org/apache/hudi/io/hfile/HFileBlockCache.java

```diff
@@ -0,0 +1,165 @@
+ * under the License.
+ */
+
+package org.apache.hudi.io.hfile;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Least Frequently Used (LFU) cache for HFile blocks to improve read performance by avoiding repeated block reads.
+ * Uses Caffeine cache with configurable size and TTL. Thread-safe for concurrent access.
+ */
+public class HFileBlockCache {
+
+  private final Cache<BlockCacheKey, HFileBlock> cache;
+
+  public HFileBlockCache(int maxCacheSize) {
+    this(maxCacheSize, 30, TimeUnit.MINUTES);
+  }
+
+  public HFileBlockCache(int maxCacheSize, long expireAfterWrite, TimeUnit timeUnit) {
+    this.cache = Caffeine.newBuilder()
+        .maximumSize(maxCacheSize)
+        .expireAfterWrite(Duration.ofMillis(timeUnit.toMillis(expireAfterWrite)))
+        .build();
+  }
+
+  /**
+   * Gets a block from cache.
+   *
+   * @param key the cache key
+   * @return cached block or null if not found
+   */
+  public HFileBlock getBlock(BlockCacheKey key) {
+    return cache.getIfPresent(key);
+  }
+
+  /**
+   * Puts a block into cache.
+   *
+   * @param key   the cache key
+   * @param block the block to cache
+   */
+  public void putBlock(BlockCacheKey key, HFileBlock block) {
+    cache.put(key, block);
+  }
+  
+  /**
+   * Gets a block from cache, or computes and caches it if not present.
+   * This method is thread-safe and prevents the "cache stampede" problem
+   * where multiple threads try to load the same value simultaneously.
+   *
+   * @param key      the cache key
+   * @param loader   callable to load the block if not in cache
+   * @return cached or newly computed block
+   * @throws Exception if the loader throws an exception
+   */
+  public HFileBlock getOrCompute(BlockCacheKey key, Callable<HFileBlock> loader) throws Exception {
+    // Caffeine uses Function instead of Callable, so we need to wrap the Callable
+    return cache.get(key, (k) -> {
+      try {
+        return loader.call();
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    });
+  }
+
+  /**
+   * Clears all cached blocks.
+   */
+  public void clear() {
+    cache.invalidateAll();
+  }
+
+  /**
+   * Gets current cache size.
+   *
+   * @return number of cached blocks
+   */
+  public long size() {
+    return cache.estimatedSize();
+  }
+
+  /**
+   * Forces cache maintenance operations like eviction.
+   * This is useful for testing to ensure consistent behavior.
+   */
+  public void cleanUp() {
+    cache.cleanUp();
+  }
+
+  /**
+   * Cache key for identifying blocks uniquely.
+   */
+  public static class BlockCacheKey {
+
+    private final String fileIdentity;
+    private final long offset;
+    private final int size;
+
+    public BlockCacheKey(long offset, int size) {
+      this(null, offset, size);
+    }
```

Comment:
Should this be removed to avoid being used for multiple files with the same offset and size which can cause wrong data being loaded from the cache?